File: leetcode/49.字母异位词分组/49.字母异位词分组.go
package main

import "sort"

func groupAnagrams(strs []string) [][]string {
	mapping := map[string][]string{}
	for _, str := range strs {
		// 转换成字节数组，并进行排序
		s := []byte(str)
		sort.Slice(s, func(i, j int) bool {
			return s[i] < s[j]
		})
		sortedStr := string(s)
		mapping[sortedStr] = append(mapping[sortedStr], str)
	}
	result := make([][]string, 0, len(mapping))
	for _, val := range mapping {
		result = append(result, val)
	}
	return result
}

File: leetcode/49.字母异位词分组/49.字母异位词分组.py
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        """
        字母异位词，相同字母，不同位置
        排序后位置也相同，用排序后字母做为map的key，原字母作为val
         - 哈希表：{排序后字符串：[原字符串]}
        再遍历哈希表的values()即可
        """
        sorted_strs = {}
        for s in strs:
            ss = ''.join(sorted(s))
            if ss in sorted_strs:
                sorted_strs[ss].append(s)
            else:
                sorted_strs[ss] = [s]

        result = []
        for val in sorted_strs.values():
            result.append(val)
        return result


if __name__ == '__main__':
    strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
    print(Solution().groupAnagrams(strs))
